Sources/Node/Utilities/Errors.swift

```diff
@@ -1,55 +1,114 @@
 @_exported import Debugging
 
-public struct NodeError: Debuggable {
-    public let node: Node?
-    public let expectation: String
-    public let path: String
+public protocol NodeError: Debuggable {}
 
-    public let type: String = NodeError.unableToConvert // replacing the enum case to identify error type
-
-    internal init<S: SchemaWrapper>(node: S?, expectation: String, indexers: [PathIndexer] = []) {
-        self.node = node.flatMap { Node($0) }
-        self.expectation = expectation
-        self.path = indexers.path()
-    }
-
-    internal init(node: Node?, expectation: String, indexers: [PathIndexer] = []) {
-        self.node = node
-        self.expectation = expectation
-        self.path = indexers.path()
-    }
-}
-
-extension NodeError {
-    public static var unableToConvert: String { return "unableToConvert" }
+public enum Error: NodeError {
+    case invalidContainer(container: String, element: String)
+    case unableToConvert(input: Node?, expectation: String, path: [PathIndexer])
 }
 
-extension NodeError: CustomStringConvertible {
-    public var description: String {
-        return "Expected \(expectation) Got: \(node ?? .null) forPath: \(path)"
-    }
-}
-
-extension NodeError {
+extension Error {
     public var identifier: String {
-        return type
+        switch self {
+        case .invalidContainer:
+            return "invalidType"
+        case .unableToConvert:
+            return "unableToConvert"
+        }
     }
 
     public var reason: String {
-        return "Unable to convert: `\(node ?? .null)` to expectation: `\(expectation)` forPath: `\(path)`"
+        switch self {
+        case .invalidContainer(let container, let element):
+            return "Unable to use container '\(container)' with element '\(element)' expected a \(NodeConvertible.self) type"
+        case .unableToConvert(let node, let expectation, let path):
+            return "Unable to convert: `\(node ?? .null)` to expectation: `\(expectation)` forPath: `\(path)`"
```

Comment:
Something like this would be much more readable:

```swift
let path = path.joined(separator: ".")
if let node = node, node != .null {
    return "Unable to convert '\(node)' to '\(expectation)' for path '\(path)'"
} else {
    return "No value found at path '\(path)', expected '\(expectation)'"
}
```

It's confusing to read `"Unable to convert: nil to expectation: Bool"`.